<!DOCTYPE html>
<html lang="en">

  <head>

  <meta charset="UTF-8">
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <!-- Meta -->
  <title>
    
      Research &middot; Dr. sc. nat. Lorenzo Gatti
    
  </title>
  <meta name="description" content="Solutions Architect interested in high performance cloud computing, data science, and taming complexity">
  <meta name="keywords" content="cloud, data analytics, business intelligence, cloud computing, aws, hpc, high performance computing, molecular evolution, computer science, evolution, dynamics, phylogenomics, trees, computational biology, bioinformatics">
  <meta name="theme-color" content="#268bd2">

  <!-- Twitter -->
  <meta name="twitter:card" content="summary" />
  <meta name="twitter:site" content="@lore_89" />
  <meta name="twitter:title" content="Dr. sc. nat." />
  <meta name="twitter:description" content="Solutions Architect interested in high performance cloud computing, data science, and taming complexity" />
  <meta property="twitter:image" content="https://lorenzogatti.me/assets/about/lorenzo3.jpg" />

  <!-- Facebook -->
  <meta property="og:site_name" content="Dr. sc. nat." />
  <meta property="og:title" content="Research" />
  <meta property="og:url" content="https://lorenzogatti.me/research/" />
  <meta property="og:image" content="https://lorenzogatti.me/assets/about/lorenzo3.jpg" />
  <meta property="og:description" content="Solutions Architect interested in high performance cloud computing, data science, and taming complexity" />
  <meta property="og:type" content="article" />

  <!-- CSS -->
  <link rel="stylesheet" href="/assets/styles.css">

  <!-- Icons -->
  <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/assets/apple-touch-icon-144-precomposed.png">
  <link rel="shortcut icon" href="/assets/favicon.ico">


  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

</head>


  <body>
  	
  		<div id="navigation" class="headroom">
  <div class="navcontrols">
    <a id="menu">Menu</a>
    <a id="close">Close</a>
  </div>
  <div class="navbar unstyle">
    <nav>
      <ul>
        <!--<li><a href="javascript:history.back()">Go Back</a></li>-->
        <li><a href="javascript:document.location.href='/';">Home</a></li>
        
        
          
            
          
        
          
        
          
            
          
        
          
        
          
        
          
        
          
        
          
            
              <li><a id="" href="/about/">About me</a></li>
            
          
        
          
            
              <li><a id=" active" href="/research/">Research</a></li>
            
          
        
          
            
              <li><a id="" href="/flying/">Flying</a></li>
            
          
        
          
            
              <li><a id="" href="/events/">Events</a></li>
            
          
        
          
            
              <li><a id="" href="/fhj/">FHJ</a></li>
            
          
        
          
            
              <li><a id="" href="/contacts/">Contact me</a></li>
            
          
        

        <li><a href="https://bitbucket.com/lorenzogatti89/"><img style="display: inline-block;vertical-align: middle;"src="/assets/bitbucket.png" height="20" width="20"></a></li>
        <li><a href="https://github.com/gattil"><img style="display: inline-block;vertical-align: middle;"src="/assets/GitHub-Mark-32px.png"height="20" width="20" ></a></li>
        <li><a href="https://devops.lorenzogatti.me/"><img style="display: inline-block;vertical-align: middle;"src="/assets/jira.png" height="20" width="20"></a></li>
      </ul>
    </nav>
  </div>
</div>

    

    <div id="sidebar" style="background-image: url(/assets/about/lorenzo3.jpg);">
  <div class="sidecontainer">
    
    <div class="sidebar-logo">
      <a href="http://lorenzogatti.me"><div class="logo" style="background-image: url(/assets/logo.png);
"></div></a>
    </div>
    
    <div class="sidebar-sticky">
      <div class="sidebar-about">
        <h2 style="line-height: 1;">
          <a href="https://lorenzogatti.me" style="line-height: 0.5;">
            <small style="font-size:medium;">Dr. sc. nat.</small><br>
            Lorenzo Gatti
          </a>
        </h2>
        <p class="lead">Solutions Architect interested in high performance cloud computing, data science, and taming complexity</p>
      </div>
      <div id="social" class="unstyle">
        <ul>
          

          
            
              <li>
                <a href="https://www.linkedin.com/in/lorenzogatti">LinkedIn</a>
              </li>
              <li class="social-separator">|</li>
            
          
            
              <li>
                <a href="https://twitter.com/lore_89">twitter</a>
              </li>
              <li class="social-separator">|</li>
            
          
            
              <li>
                <a href="https://www.researchgate.net/profile/Lorenzo_Gatti">ResearchGate</a>
              </li>
              <li class="social-separator">|</li>
            
          
            
              <li>
                <a href="https://uzh.academia.edu/LorenzoGatti">academica</a>
              </li>
              <li class="social-separator">|</li>
            
          
        </ul>
      </div>
    </div>
  </div>
</div>


    <div class="content container">
		<article class="page">
  <h1 class="page-title">Research</h1>
  <p><br />
I studied the evolutionary molecular and epidemiological dynamics of viral diseases affecting the human population on a global scale, and how to implement a new class of evolutionary models which account for incorporing insertion and deletions events into phylogenetic reconstruction.</p>

<h3 id="algorithms-for-optimising-binary-tree-topologies">Algorithms for optimising binary tree topologies</h3>

<p>The tree topology is a multidimensional parameter that requires a specific class of algorithms for its optimization. Since exploring the tree space is a highly computationally demanding step, we use the Unified Tree Search Heuristics C++ Library (TSHLib), a phylogenetic library optimised for fast tree-search that supports multi-core parallelization.  We tuned the tree-search acting at several levels: proposing algorithm, tree-space coverage, node seeding methods. To compute a tree rearrangement we implemented two <em>proposing algorithms</em>: a fast <em>node-swap</em> and a more computationally demanding <em>node-cut</em> methods. Both these two methological classes are well established in the community, and their computational limits have been widely studied.</p>

<table>
  <thead>
    <tr>
      <th><img src="/assets/research/TreeHeuristics.png" alt="steam-fish-1" width="800" /></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>The three standard tree rearrangement operations (NNI, SPR, and TBR).</strong> The three basic tree rearrangement operations (NNI, SPR, and TBR). In SPR and TBR all pairs of ``circled’’ branches among the two subtrees will be connected (dashed lines), except the two filled circles to each other, since this yields the full tree again</td>
    </tr>
  </tbody>
</table>

<p>In our implementation, at the beginning of the tree-search cycle, the tree-search algorithm proceeds by proposing a set of rearrangements on the current tree. The algorithm searches for all the possible non-redundant rearrangements within a radius distance \(d\) from either all the available nodes on the tree-topology <strong>(greedy)</strong>, or from a random subset <strong>(random-hillclimbing)</strong>. The radius distance \(d\) is automatically set such as NNI moves will target only the nearest neighbours nodes and SPR moves the remaining far more distant ones, respectively. The likelihood of every new tree is re-computed using a conservative approach: only the likelhood matrices pertaining to the nodes upwards (towards the root) from where the rearrangement occurred will have to be re-computed. In this manner, the re-computation is restricted only to those likelhood matrices where it is strictly necessary, and the likelhood matrices of the nodes not included in this set are fully recycled. The computation occurring for an NNI move is a special case of an SPR move. The process is shown in details in the figure below.</p>

<table>
  <thead>
    <tr>
      <th><img src="/assets/research/setArecomputation_on_spr.png" alt="steam-fish-1" width="800" /></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>Recomputing the multi-set of insertion events during the tree-search optimisation.</strong> For a rooted tree topology there are multiple evolutionary scenarios which explain the pattern of gaps and characters observed at a particular multi-sequence alignment (MSA) site. The figure illustrates the re-computation of the multi-set of insertion events during tree-search optimisation for a global tree rearrangement (SPR move). This tree is rooted on node \((v_{\omega})\), and has 16 internal nodes \((v_a,v_b,v_c,v_d,v_e,v_f,v_g,v_h,v_z,v_y,v_x,v_x, v_w, v_v, v_u, v_t, v_z)\) and 18 leaves \((v_{1,\cdots,}v_{18})\). The tree represents the phylogeny of a single site (or column of the MSA), \(s_i=\{v_1:(T);v_2:(T);v_3:(T);v_4:(C);v_5:(A);v_{6,\cdots,18}:(-)\}\). At the indicated site, the subtree indicated in bold is pruned at node \(v_z\) and regrafted between nodes \(v_t\) and \(v_i\) at a distance \(S\). Nodes \(v_a\) (orphan) and \(v_y\) (step-parent) are reconnected, and at the regraft site, the connection between nodes \(v_t\) and \(v_i\) is broken. Nodes in the rearrangement path (coloured in red) \(\{v_z,v_y,v_x,v_w,v_t,\ldots,v_u,v_v,\ldots,\Omega\}\), with \(\Omega\) being the root of the tree, must update the multi-set of insertion events \((i)\), here indicated by triangles. Before tree rearrangement (blue triangles) the multi-set of insertion events leading to the observed characters is: \(i_A = \{i_{A_0}; i_{A_1}; i_{A_2}\}\), and after tree rearrangement (dashed triangles) is updated into: \(i_B = \{i_{B_0}\}\)</td>
    </tr>
  </tbody>
</table>

<p>We tested the soundness of our implementation’s tree-search algorithms by comparing the tree topologies inferred by Castor with those inferred by other softwares (i.e. codonPhyML). With increasing number of sites in the alignments both Castor and CodonPhyML eventually find the true tree on average. It is important here to notice that the variance of the estimates inferred by Castor and CodonPhyML is due to the distance-based initial trees. Lastly, we repeated this test, but this time inferring the tree topology under the true indel-aware substitution model (K80+PIP). We were able to recover the true value for all the numerical parameters and the true tree topology.</p>

<table>
  <thead>
    <tr>
      <th><img src="/assets/research/sanity_3_1.png" alt="steam-fish-1" width="800" /></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>Benchmark tree-search algorithms: Comparison of tree-search accuracy under indel-aware substitution models for increasing data volumes using Castor and codonPhyML.</strong> A) Value estimates of the kappa parameter inferred from data set A2 (generated under K80) under the traditional substitution model K80 implemented in Castor and codonPhyML (denoted Castor and PhyML in the inset, respectively). The initial bioNJ tree topology is optimized using NNI moves. The boxes and whiskers are computed from 100 estimates per data set (shown on x-axis). Inferences are obtained from data of increasing expected length (100, 1000, 1000 sites). B) is analogous to A), with only the log-likelihoods estimates. C) and D) are analogous to A), but showing the nRF measure between inferred trees and the true tree used to generate the data, and total tree lengths inferred by each program.</td>
    </tr>
  </tbody>
</table>

<p>After proving that our implementation is sound, we moved forward with testing the asymptotic convergence of the ML estimates under indel-aware substitution models. The figure below shows that with increasing number of sites in the alignments, both the numerical parameter estimates and the tree topology asymptotically converge to their true values (panels I-L for 60 taxa). As expected, the variance around all the numerical parameter estimates diminishes with the increasing the alignment sites, as well as the tree topology distance from the true tree. It is worth noticing that already for small datasets, using a <strong>mixed tree-search strategy</strong>, the estimates of both numerical parameters and tree topology were already close to their true values. We therefore suggest the use of this option in real case studies.</p>

<table>
  <thead>
    <tr>
      <th><img src="/assets/research/asymptotic_figure.png" alt="steam-fish-1" width="800" /></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><strong>Asymptotic convergence of numerical model parameters and tree topology under the true model using different tree-search strategies.</strong> An initial bioNJ tree topology is optimized using either NNI, SPR, or Mixed moves (denoted in the inset as nni, spr, mixed), while four parameters are monitored: the substitution parameter of the K80 model – \(\kappa\) –; the intensity of the PIP process – \(I\) – ; the measure between inferred trees and the true tree used to generate the data – <em>nRF</em> –; and the total tree length. The boxes and whiskers are computed from 100 estimates per data set (shown on x-axis). Inferences are obtained from data of increasing expected length (100, 1000, 1000 sites).</td>
    </tr>
  </tbody>
</table>

<p>In Castor – a new phylogenetic Maximum Likelihood inference software which adds the PIP model as an extension to the most common traditional substitution models – we incorporated well-established tree inference methods (see above) into a framework for indel-aware phylogenetic inference. This software solution is based on the Bio++ Bioinformatics Libraries (Dutheil and Bigot, 2018) and inspired by the bppML module of the BppSuite programs for phylogenetic and sequence analysis. To boost computational time and memory allocation performances by better harnessing of modern CPU architectures, we built the framework, algorithms, and data-structures using C++ Object Oriented Programming (OOP) design patterns, taking advantage of code vectorisation and multithread paradigms.</p>

<hr />

<h3 id="grants">Grants</h3>

<ul>
  <li>1 Feb 2018 - 31 Jan 2019: University of Zürich - Forschungskredit Candoc grant no. K-74409-01-01</li>
  <li>1 July 2015: UZH-ETH MLS Travel grant</li>
</ul>

<hr />

<h3 id="teaching">Teaching</h3>
<ul>
  <li><a href="http://lorenzogatti.me/2017-ACGT-Tutorials/">GENECO course - Computational Molecular Evolution: Detecting Positive Selection -  at University of Lund</a></li>
  <li><a href="http://gattil.github.io/2016_FiPS_Tutorials">Frontiers in Plant Sciences 2016: Phd course in protein-coding evolution and detecting natural selection</a></li>
  <li><a href="http://gemp2015.lifeinnumbers.ch/program/">ZHAW-SIB Summer School GEMP 2015: Genomics and Evolution of Microbial Pathogens</a></li>
</ul>

<hr />

<h3 id="publications">Publications</h3>
<ol>
  <li>
    <p><strong>Lorenzo Gatti</strong>, Jitao David Zhang, Maria Anisimova, Martin Schutten, Albert Osterhaus, Erhard van der Vries,  Cross-reactive immunity drives global oscillation and opposed alternation patterns of seasonal influenza A viruses, BioRxiv (2017) <a href="https://doi.org/10.1101/226613">doi:10.1101/bioRxiv/226613</a></p>
  </li>
  <li>
    <p>Davide Heller, Andreas Hoppe, Simon Restrepo, <strong>Lorenzo Gatti</strong>, Alexander L. Tournier, Nicolas Tapon, Konrad Basler, and Yanlan Mao (2016). EpiTools: An Open-Source Image Analysis Toolkit for Quantifying Epithelial Growth Dynamics. Developmental Cell 36 (1) (January): 103–116. <a href="http://dx.doi.org/10.1016/j.devcel.2015.12.012">doi:10.1016/j.devcel.2015.12.012</a></p>
  </li>
  <li>
    <p>SIB members, including: Anisimova, Maria; <strong>Gatti, Lorenzo</strong>; Gil, Manuel (2015). The SIB Swiss Institute of Bioinformatics’ resources: focus on curated databases. In: Nucleic Acids Research. Oxford: Oxford University Press.
<a href="http://nar.oxfordjournals.org/content/early/2015/11/27/nar.gkv1310.abstract">doi: 10.1093/nar/gkv1310</a>.</p>
  </li>
</ol>

<hr />

<h3 id="software">Software</h3>

<ul>
  <li><a href="https://bitbucket.org/lorenzogatti89/castor">Castor</a>. Indel-Aware Phylogenetic Tree Inference Software</li>
  <li><a href="https://bitbucket.org/lorenzogatti89/castor-aws">Castor AWS</a>. Auto-scaling On-Demand Indel-Aware Phylogenetic Tree Inference on the public cloud</li>
  <li><a href="https://bitbucket.org/lorenzogatti89/tshlib">Tree Search Heuristic Library</a>. The Unified Tree-Search Heuristic C++ Library (TSH-LIB) is a new unified implementation of binary tree topological searching engines for phylogenetic analyses.</li>
  <li><a href="https://bitbucket.org/lorenzogatti89/alist/src/master/">AliSt</a>. ALIgnment STatistic, a program to compute indel presence statistics on Multi-Sequence Alignments.</li>
  <li><a href="https://github.com/gattil/netxDiscreteDynamics">NetXDiscreteDynamics</a>. Probabilistic framework for discrete evolutionary dynamic</li>
  <li><a href="https://github.com/gattil/phylo-tools">Phylo-Tools</a>. Phylogenetic tree manipulation toolbox</li>
  <li><a href="https://epitools.github.io/wiki/">Epitools</a>. Image processing software for Epithelial Tissues</li>
</ul>

</article>


	<footer>The <a href="http://skye.connorbaer.io">Skye</a> theme <a href="https://github.com/connor-baer/skye">v2.0.0</a>. &copy; 2024. All rights reserved. <a href="https://github.com/gattil/edit/gh-pages/02_research/index.md">Edit this page</a>.</footer>

    
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-73302940-1', 'auto');
  ga('require', 'linkid');
  ga('send', 'pageview');

</script>

<script type="text/javascript" src="/assets/scripts.js"></script> 
  </body>
</html>
